client/state/selectors/should-display-app-banner.ts

```diff
@@ -0,0 +1,68 @@
+import { createSelector } from '@automattic/state-utils';
+import { isMobile } from '@automattic/viewport';
+import { includes } from 'lodash';
+import {
+	APP_BANNER_DISMISS_TIMES_PREFERENCE,
+	ALLOWED_SECTIONS,
+	isDismissed,
+	getCurrentSection,
+} from 'calypso/blocks/app-banner/utils';
+import { isWpMobileApp } from 'calypso/lib/mobile-app';
+import { getPreference, hasReceivedRemotePreferences } from 'calypso/state/preferences/selectors';
+import isNotificationsOpen from 'calypso/state/selectors/is-notifications-open';
+import { shouldDisplayTosUpdateBanner } from 'calypso/state/selectors/should-display-tos-update-banner';
+import { getSectionName, appBannerIsEnabled } from 'calypso/state/ui/selectors';
+import { AppState } from 'calypso/types';
+
+/**
+ * Returns true if the App Banner should be displayed
+ *
+ * @param {object} state Global state tree
+ * @returns {boolean} True if App Banner is visible
+ */
+export const shouldDisplayAppBanner = ( state: AppState ): boolean | undefined => {
+	// The ToS update banner is displayed in the same position as the mobile app banner. Since the ToS update
+	// has higher priority, we repress all other non-essential sticky banners if the ToS update banner needs to
+	// be displayed.
+	if ( shouldDisplayTosUpdateBanner( state ) ) {
+		return false;
+	}
+
+	// In some cases such as error we want to hide the app banner completely.
+	if ( ! appBannerIsEnabled( state ) ) {
+		return false;
+	}
+
+	if ( ! hasReceivedRemotePreferences( state ) ) {
+		return false;
+	}
+
+	if (
+		typeof window !== 'undefined' &&
+		window.sessionStorage.getItem( 'wpcom_signup_complete_show_draft_post_modal' )
+	) {
+		return false;
+	}
```

Comment:
The `shouldDisplayAppBanner` selector is called in the main `Layout` component every time it's re-rendered, which is very often. Reading from local storage on every Redux state change is very inefficient. Selectors should only read from `state`, not from any other data source.

Reply:
Thanks, @jsnajdr, that is a valid point. I opened [a new issue](https://github.com/Automattic/wp-calypso/issues/60129) to solve this.